packages/matter.js/src/device/CodeModel.ts

```diff
@@ -0,0 +1,232 @@
+            firstChar = firstChar.toUpperCase();
+        } else {
+            firstChar = firstChar.toLowerCase();
+            didFirst = true;
+        }
+        return `${firstChar}${piece.slice(1).toLowerCase()}`;
+    }).join("");
+}
+
+class LeafModel<T> {
+    constructor(public name: string, public definition: T, private parentSource: string) { }
+    get typeName() { return camelize(this.name); }
+    get fieldName() { return camelize(this.name, false); }
+    get source() { return `${this.parentSource}.${camelize(this.fieldName, false)}` }
+}
+
+class LeafModels<T, ST> extends Array<T> {
+    constructor(definitions: { [key: string]: ST }, Cons: new (name: string, definition: ST, cluster: string) => T, cluster: string) {
+        // Use the name of the leaf class minus "Model" as the path in the cluster
+        const type = Cons.name.slice(0, Cons.name.length - 5).toLowerCase();
+        super(...Object.entries(definitions).map(([k, v]) => new Cons(k, v, `${cluster}.${type}s`)));
+    }
+}
+
+export class AttributeModel extends LeafModel<clusterExports.Attribute<any>> {
+    get writable() { return this.definition.writable; }
+    get id() { return this.definition.id; }
+    get schema() { return this.definition.schema; }
+    get optional() { return this.definition.optional; }
+    get default() { return this.definition.default }
+    get handler() { return `on${this.typeName}Change` }
+    get getter() { return `get${this.typeName}` }
+    get fixed() { return false; /* TODO - update when metadata lands */ }
+}
+
+export class CommandModel extends LeafModel<clusterExports.Command<any, any>> {
+    get requestSchema() { return this.definition.requestSchema; }
+    get responseSchema() { return this.definition.responseSchema; }
+    get invoker() { return `send${this.typeName}`; }
+    get handler() { return `on${this.typeName}`; }
+}
+
+export class EventModel extends LeafModel<clusterExports.Event<any>> {
+    get schema() { return this.definition.schema; }
+    get handler() { return `on${this.typeName}`; }
+    get invoker() { return `trigger${this.typeName}`; };
+}
+
+export type UntypedHandlers = { [key: string]: (...any: any[]) => any };
+type UntypedHandlerFactory = () => UntypedHandlers;
+
+function findDefaultHandlers(cluster: ClusterModel): UntypedHandlers {
+    let handlers = <UntypedHandlerFactory>(<any>clusterExports)[`${cluster.name}ClusterHandler`];
+    if (handlers && typeof handlers == "function") return handlers();
+
+    // Some of the handlers don't have "Cluster" in their name
+    handlers = <UntypedHandlerFactory>(<any>clusterExports)[`${cluster.name}Handler`];
+    if (handlers && typeof handlers == "function") return handlers();
+
+    return {};
+}
+
+function defaultState(cluster: ClusterModel) {
+    const result = <{ [key: string]: any }>{};
+
+    cluster.attributes.forEach((attr) => {
+        if (attr.default !== undefined) {
+            result[attr.name] = attr.default;
```

Comment:
see above ... ideally do not use that ... just use "getLocal" ... Reason: There is also maybe storage involved and so the initial value and default value are not relevant at all

Reply:
I did this because the handlers have createDefaultFoo with default values in them.  Other places only have the bare handlers.  In order to make it work everywhere I targeted the least-common denominator.

Maybe there could be a generic `createDefaultClusterServer(cluster, attributeInitialValues, handlers)` that does the right thing?  Might be able to slim down the server files if we set the defaults from the schema.

I looked at a handful of the createDefaultFoo, maybe they could all be replaced by this?  And then use FooClusterHandler for further specialization since it's already a factory?

Reply:
I also had such an idea already